### DISCUSSION TITLE
Error downloading past Peloton workouts - Peloton app subscription only

### DISCUSSION THREAD

**Discussion by epaksoy** [philosowaffle/peloton-to-garmin#341] [General]
I am new to this. I tried to post this earlier today, but I do not see it in the discussions. Sorry if this is a duplicate.

I just discovered this app and thought it might solve my problem. I have an Echelon bike and a Peloton app subscription. I am able to upload Peloton and Zwift ride fit file data to Garmin. I am hoping that this app can help me upload my Peloton strength workouts as well.

I downloaded, configured and ran P2G 3.0.3 on my Windows 10 laptop. Initially I had Peloton.NumWorkoutsToDownload: 1,  and Garmin.Upload: "false", so it only downloaded my last core workout from yesterday as a fit file. I was also able to upload it to Garmin (though, no details - just the class name and the time info). I updated those to 1000 and "true" respectively, and re-ran. The connection works, and I can see my Peloton profile info and the number of workouts in the log. But I got some error messages and no downloads. 

Notably:
2022-09-25 11:08:46.739 -05:00 [ERR] Failed to fetch UserData from Peloton. FTP info may be missing for certain non-class workout types (Just Ride).
Flurl.Http.FlurlParsingException: Response could not be deserialized to JSON: GET https://api.onepeloton.com/api/me
 ---> Newtonsoft.Json.JsonSerializationException: Error converting value {null} to type 'Common.Dto.Peloton.CyclingFtpSource'. Path 'cycling_ftp_source', line 1, position 5490.

and many of these:
2022-09-25 11:08:47.460 -05:00 [ERR] Http Call Failed. null 

after which:
Failed to download workouts from Peloton.
Flurl.Http.FlurlHttpException: Call failed. The SSL connection could not be established, see inner exception: GET https://api.onepeloton.com/api/workout/7d9a03d4a0854db4a03778a4cfff80ef/performance_graph?every_n=1
 ---> System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.

I am attaching the config (fake credentials) and log. I hope you can point me in the right direction.

Thanks in advance.

[log20220925.txt](https://github.com/philosowaffle/peloton-to-garmin/files/9641549/log20220925.txt)
[configuration.local - Copy.txt](https://github.com/philosowaffle/peloton-to-garmin/files/9641550/configuration.local.-.Copy.txt)

**Comment by philosowaffle** [comment]
Hi @epaksoy, thank you for logging this and sharing all the logs/configs!  From your logs, I found two problems in the code that need to addressed and have logged bugs for these #342 and #343 .

I'll post back here and let you know when the fixes are done.

Regarding:
> I am able to upload Peloton and Zwift ride fit file data to Garmin. I am hoping that this app can help me upload my Peloton strength workouts as well.

If your goal is to use P2G only for syncing Strength workouts, you may find it beneficial (and speed up your sync) to exclude all other workout types from P2G.  You can configure that setting in the [Peloton config section](https://philosowaffle.github.io/peloton-to-garmin/configuration/json.html#peloton-config) by setting the [ExcludeWorkoutTypes](https://philosowaffle.github.io/peloton-to-garmin/configuration/json.html#exclude-workout-types) to a comma separated list of workout types you'd like P2G to ignore.

**Comment by epaksoy** [comment]
Thank you very much for the quick response. I will be sure to exclude workout types when I run it again. I’ll be checking on the status of the bugs before I do that.On Sep 25, 2022, at 3:46 PM, philosowaffle ***@***.***> wrote:﻿
Hi @epaksoy, thank you for logging this and sharing all the logs/configs!  From your logs, I found two problems in the code that need to addressed and have logged bugs for these #342 and #343 .
I'll post back here and let you know when the fixes are done.
Regarding:

I am able to upload Peloton and Zwift ride fit file data to Garmin. I am hoping that this app can help me upload my Peloton strength workouts as well.

If your goal is to use P2G only for syncing Strength workouts, you may find it beneficial (and speed up your sync) to exclude all other workout types from P2G.  You can configure that setting in the Peloton config section by setting the ExcludeWorkoutTypes to a comma separated list of workout types you'd like P2G to ignore.

**Comment by philosowaffle** [comment]
I was able to fix one of the issues you encountered, but unfortunately I'm having trouble reproducing the second one on my machine:

```
Flurl.Http.FlurlHttpException: Call failed. The SSL connection could not be established, see inner exception: GET https://api.onepeloton.com/api/workout/7d9a03d4a0854db4a03778a4cfff80ef/performance_graph?every_n=1
---> System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
```

If you have time I wonder if you could try the following for me and share your findings?

1. [Click here](https://github.com/philosowaffle/peloton-to-garmin/suites/8573629351/artifacts/383649223) to download the latest windows exe, it has some additional logging in it that may help me (plus the fix for the first error in your original message)
2. Temporarily disable Garmin Upload, since this problem is isolated to just fetching the workouts from Peloton don't want to spam Garmin until we've got the first part working.
3. Try syncing 1 workout again as a sanity check, then try 50, then 150, etc.  Trying to find the threshold where it goes from working to not working.

**Reply by epaksoy** [reply]
Thank you very much for working on this. The first issue appears to be resolved. 

As for the second, I think I narrowed down the breaking point to around 210 workouts on my PC.
210 worked. 220 also worked, at least initially. But when I reran 220 immediately afterwards, it gave errors.
It is also worth noting that in all of these cases only 200 files were downloaded.

Here are the logs for the successful 210 download and the unsuccessful 220 download.
[Output-10032022-0344-220downloads-unsuccessful.txt](https://github.com/philosowaffle/peloton-to-garmin/files/9700868/Output-10032022-0344-220downloads-unsuccessful.txt)
[Output-10032022-0343-210downloads-successful.txt](https://github.com/philosowaffle/peloton-to-garmin/files/9700869/Output-10032022-0343-210downloads-successful.txt)

**Comment by philosowaffle** [comment]
@epaksoy sorry for the long delay, life's been busy. But I've got a new build for you to download and try:

https://github.com/philosowaffle/peloton-to-garmin/suites/8790363058/artifacts/399358544

**Reply by epaksoy** [reply]
Thank you so much. It worked! 

I first tried just downloading 1000 files. It successfully downloaded fit files for all 835 non cycling activities. 
Then I deleted the fit files, set Garmin upload to "true",  and ran it again. It first downloaded all of them and it appears that it uploaded all of them successfully!

I would like to know how I should use it going forward. What should my settings be? Does it detect duplicates on upload to Garmin? Do I need to clear the output/fit folder? Do I run it manually everyday? I see that it sleeps in the background (default 24 hrs). Do I restart it whenever I reboot? 

I do something on Peloton on most days. Sometimes as many as 5 (short) activities in a day.

Thanks again!

**Reply by philosowaffle** [reply]
Yay! Glad to hear it worked!

The P2G Windows exe can be used in two ways, and so it'll just be personal preference based on what you'd like to do.

Below are some sample configs, but there is more thorough documentation on each and every field available on the [docs site](https://philosowaffle.github.io/peloton-to-garmin/configuration/json.html#app-config).

No matter which direction you choose, it is safe to go ahead and delete the `output/fit` folder.  This folder is helpful when debugging, for generating a personal backup of the files, or if a person wishes to manually upload the files somewhere other than Garmin.  For your use case, they are no longer needed, and in the sample configs below I have shown how to disable P2G saving copies to this folder going forward.

### Manually run as needed

Whenever you want to sync, just manually run the P2G exe, then exit when finished.  With the below config file:

1. Each time you want to sync, the P2G terminal will ask you how many workouts it should grab, (it always grabs the most recent workouts first)
2. You will not need to worry about duplicates, Garmin will reject any workouts that have already been uploaded
3. `SaveLocalCopy: false` - ensures that you won't have an ever-growing Output folder, P2G will clean up as it goes

Note that I have truncated it to just the fields that need modification:

```
{
  "App": {
     ....
    "EnablePolling": false, 
    "CheckForUpdates": true,
    "CloseWindowOnFinish": true // optional based on personal preference, P2G can exit automatically when done if desired
     ....
  },

  "Format": {
    .....
    "SaveLocalCopy": false,
   .....
  },

  "Peloton": {
    .....
    "NumWorkoutsToDownload": 0,
    "ExcludeWorkoutTypes": [ list of non-strength exercises]
  },

..... truncated, no changes to the rest
}
```

### Auto-Sync on a schedule

You'll launch P2G once and it will sync on a schedule. On Windows, you'll have to minimize the P2G terminal once its running (it does not truly run in the background).  Exiting out of the P2G terminal does kill the program.  And if your computer reboots, you will have to remember to restart P2G.

With the below config file:

1. P2G will sync once every `86400` seconds (24hrs)
2. It will automatically attempt to sync the most recent `NumWorkoutsToDownload`
3. You will not need to worry about duplicates, Garmin will reject any workouts that have already been uploaded
4. `SaveLocalCopy: false` - ensures that you won't have an ever-growing Output folder, P2G will clean up as it goes

```
{
  "App": {
     ....
    "EnablePolling": true, 
    "PollingIntervalSeconds": 86400,
    "CheckForUpdates": true
     ....
  },

  "Format": {
    .....
    "SaveLocalCopy": false,
   .....
  },

  "Peloton": {
    .....
    "NumWorkoutsToDownload": 5,
    "ExcludeWorkoutTypes": [ list of non-strength exercises]
  },

..... truncated, no changes to the rest
}
```

**Reply by epaksoy** [reply]
Thank you so much for the detailed information. I will plan to do it manually for now. I already do that with my Peloton and Zwift cycling workouts using an iPhone app called "QZ - qdomyos-zwift", which links my Echelon bike to those two apps in "closed loop" mode with automatic resistance adjustment. Thankful for the support from two great independent developers.

**Reply by epaksoy** [reply]
Here is something I ran into today. I took three classes and then I ran P2G. Only two uploaded. I checked and it looks like I had taken the third one back in June. I am wondering if that's why it didn't upload. How are duplicates verified? Based on the class name? Do I understand correctly that the verification occurs on the Garmin side?

**Reply by philosowaffle** [reply]
The verification occurs Garmin side and my understanding is this is usually based on a matching timestamp not class name.  Did the log output give any additional info?

**Reply by epaksoy** [reply]
Yes. Mystery solved, I think.

The logs form yesterday did not show the third workout being downloaded from Peloton (even though it was appearing in my Peloton workout history). I tried again this morning, and it was successfully downloaded from Peloton and uploaded to Garmin. I guess I may need to wait a little bit longer for things to settle in Peloton before I run P2G.

Thanks again!
